lib/stages/Transpose.cpp

```diff
@@ -0,0 +1,231 @@
+    std::string ds_id_str;
+
+    size_t frame_size = 0;
+
+    // wait for a non-empty frame to get dataset ID from
+    uint32_t num_empty_skip = 0;
+    while (true) {
+        // Wait for a frame in the input buffer in order to get the dataset ID
+        if ((wait_for_full_frame(in_buf, unique_name.c_str(), frame_id)) == nullptr) {
+            return;
+        }
+
+        // Get the frame size to publish metrics later
+        uint64_t fpga_seq_total;
+        dset_id_t frame_ds_id;
+        std::tie(frame_size, fpga_seq_total, frame_ds_id) = get_frame_data();
+
+        // If the frame is empty, release the buffer and continue
+        if (fpga_seq_total == 0 && frame_ds_id == dset_id_t::null) {
+            DEBUG("Got empty frame ({:d}).", frame_id);
+            mark_frame_empty(in_buf, unique_name.c_str(), frame_id++);
+            num_empty_skip++;
+        } else {
+            ds_id = frame_ds_id;
+            break;
+        }
+    }
+
+    if (num_empty_skip > 0) {
+        INFO("Found {:d} empty frames at the start of the file.", num_empty_skip);
+    }
+
+    if (!get_dataset_state(ds_id)) {
+        FATAL_ERROR("Couldn't find ancestor of dataset {}. "
+                    "Make sure there is a stage upstream in the config, that adds the dataset "
+                    "states.\nExiting...",
+                    ds_id);
+        return;
+    }
+
+    // Get the original dataset ID (before adding time axis)
+    base_ds_id = base_dset(ds_id);
+
+    found_flags = std::vector<bool>(write_t, false);
+
+    // Create HDF5 file
+    create_hdf5_file();
+
+    // TODO: it seems like this should be a Counter?
+    auto& transposed_bytes_metric =
+        Metrics::instance().add_gauge("kotekan_transpose_data_transposed_bytes", unique_name);
+
+    while (!stop_thread) {
+        if (num_empty_skip > 0) {
+            // Write out empty frames that were skipped at start
+            // All arrays are initialised to zero, so we just need to move through them
+            num_empty_skip--;
+        } else {
+            // Wait for a full frame in the input buffer
+            if ((wait_for_full_frame(in_buf, unique_name.c_str(), frame_id)) == nullptr) {
+                break;
+            }
+
+            // Collect frames until a chunk is filled
+            // Time-transpose as frames come in
+            // Fastest varying is time (needs to be consistent with reader!)
+            copy_frame_data(fi, ti);
+
+            auto [frame_size, fpga_seq_total, frame_ds_id] = get_frame_data();
+            (void)frame_size;
+
+            // Parse the dataset ID
+            if (fpga_seq_total == 0 && frame_ds_id == dset_id_t::null) {
+                DEBUG("Got an empty frame.");
+                // Empty frames have a null dataset ID
+                ds_id_str = fmt::format("{}", dset_id_t::null);
+            } else if (frame_ds_id != ds_id) {
+                // TODO assuming that dataset ID changes here never change dataset dimensions
```

Comment:
That can essentially be tested with the fingerprint, although it's not clear it's worth doing